File: android/app/src/main/java/com/nomisafe/falldetection/FallDetectionModule.java
package com.nomisafe.falldetection;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.os.Build;
import androidx.core.content.ContextCompat;
import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.modules.core.PermissionAwareActivity;
import com.facebook.react.modules.core.PermissionListener;

public class FallDetectionModule extends ReactContextBaseJavaModule implements PermissionListener {
    private static ReactApplicationContext reactContext;
    private int listenerCount = 0;
    private Promise permissionPromise;
    private static final int LOCATION_PERMISSION_REQUEST_CODE = 1001;

    FallDetectionModule(ReactApplicationContext context) {
        super(context);
        reactContext = context;
        // Set the React context for the service to send events
        FallDetectionService.setReactContext(context);
    }

    @Override
    public String getName() {
        return "FallDetectionModule";
    }

    // Required for NativeEventEmitter
    @ReactMethod
    public void addListener(String eventName) {
        listenerCount++;
    }

    // Required for NativeEventEmitter
    @ReactMethod
    public void removeListeners(int count) {
        listenerCount -= count;
        if (listenerCount < 0) listenerCount = 0;
    }

    @ReactMethod
    public void requestLocationPermission(Promise promise) {
        permissionPromise = promise;
        
        // Check if permissions are already granted
        if (ContextCompat.checkSelfPermission(reactContext, Manifest.permission.ACCESS_FINE_LOCATION) 
                == PackageManager.PERMISSION_GRANTED) {
            promise.resolve(true);
            return;
        }
        
        // Request permissions
        PermissionAwareActivity activity = (PermissionAwareActivity) getCurrentActivity();
        if (activity == null) {
            promise.reject("NO_ACTIVITY", "Activity is null");
            return;
        }
        
        String[] permissions = {
            Manifest.permission.ACCESS_FINE_LOCATION,
            Manifest.permission.ACCESS_COARSE_LOCATION
        };
        
        activity.requestPermissions(permissions, LOCATION_PERMISSION_REQUEST_CODE, this);
    }
    
    @ReactMethod
    public void checkLocationPermission(Promise promise) {
        boolean granted = ContextCompat.checkSelfPermission(reactContext, Manifest.permission.ACCESS_FINE_LOCATION) 
                == PackageManager.PERMISSION_GRANTED;
        promise.resolve(granted);
    }

    @Override
    public boolean onRequestPermissionsResult(int requestCode, String[] permissions, int[] grantResults) {
        if (requestCode == LOCATION_PERMISSION_REQUEST_CODE && permissionPromise != null) {
            boolean granted = grantResults.length > 0 && grantResults[0] == PackageManager.PERMISSION_GRANTED;
            permissionPromise.resolve(granted);
            permissionPromise = null;
            return true;
        }
        return false;
    }

    @ReactMethod
    public void startService() {
        // Update React context in case it changed
        FallDetectionService.setReactContext(reactContext);
        Intent serviceIntent = new Intent(reactContext, FallDetectionService.class);
        reactContext.startService(serviceIntent);
        
        // Check for any pending SOS that failed to send
        FallDetectionService service = FallDetectionService.getInstance();
        if (service != null) {
            service.retryPendingSOS();
        }
    }

    @ReactMethod
    public void stopService() {
        Intent serviceIntent = new Intent(reactContext, FallDetectionService.class);
        reactContext.stopService(serviceIntent);
    }

    @ReactMethod
    public void cancelSOS() {
        // Cancel the SOS from React Native by calling the service's cancelSOS method
        FallDetectionService service = FallDetectionService.getInstance();
        if (service != null) {
            service.cancelSOS();
        } else {
            // Fallback: set flags directly if service instance not available
            FallDetectionService.sosCancelled = true;
            FallDetectionService.sosTimerActive = false;
        }
    }
    
    @ReactMethod
    public void retryPendingSOS() {
        FallDetectionService service = FallDetectionService.getInstance();
        if (service != null) {
            service.retryPendingSOS();
        }
    }
}

File: android/app/src/main/java/com/nomisafe/falldetection/FallDetectionModule.kt
package com.nomisafe.falldetection

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.os.Build
import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule

class FallDetectionModule(reactContext: ReactApplicationContext) : 
    ReactContextBaseJavaModule(reactContext) {
    
    private var fallDetectionReceiver: BroadcastReceiver? = null
    
    override fun getName() = "FallDetection"
    
    init {
        registerBroadcastReceiver()
    }
    
    @ReactMethod
    fun startMonitoring(promise: Promise) {
        try {
            val intent = Intent(reactApplicationContext, FallDetectionService::class.java)
            
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                reactApplicationContext.startForegroundService(intent)
            } else {
                reactApplicationContext.startService(intent)
            }
            
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("START_ERROR", "Failed to start monitoring: ${e.message}")
        }
    }
    
    @ReactMethod
    fun stopMonitoring(promise: Promise) {
        try {
            val intent = Intent(reactApplicationContext, FallDetectionService::class.java)
            reactApplicationContext.stopService(intent)
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("STOP_ERROR", "Failed to stop monitoring: ${e.message}")
        }
    }
    
    @ReactMethod
    fun cancelAlert(promise: Promise) {
        try {
            val intent = Intent(reactApplicationContext, FallDetectionService::class.java).apply {
                action = FallDetectionService.ACTION_CANCEL_ALERT
            }
            reactApplicationContext.startService(intent)
            promise.resolve(true)
        } catch (e: Exception) {
            promise.reject("CANCEL_ERROR", "Failed to cancel alert: ${e.message}")
        }
    }
    
    private fun registerBroadcastReceiver() {
        fallDetectionReceiver = object : BroadcastReceiver() {
            override fun onReceive(context: Context?, intent: Intent?) {
                when (intent?.action) {
                    "com.nomisafe.FALL_DETECTED" -> {
                        val params = Arguments.createMap().apply {
                            putDouble("impactG", intent.getFloatExtra("impactG", 0f).toDouble())
                            putDouble("timestamp", intent.getLongExtra("timestamp", 0L).toDouble())
                        }
                        sendEvent("onFallDetected", params)
                    }
                    "com.nomisafe.FALL_CANCELLED" -> {
                        sendEvent("onFallCancelled", null)
                    }
                    "com.nomisafe.SEND_EMERGENCY_ALERTS" -> {
                        sendEvent("onSendEmergencyAlerts", null)
                    }
                }
            }
        }
        
        val filter = IntentFilter().apply {
            addAction("com.nomisafe.FALL_DETECTED")
            addAction("com.nomisafe.FALL_CANCELLED")
            addAction("com.nomisafe.SEND_EMERGENCY_ALERTS")
        }
        
        reactApplicationContext.registerReceiver(fallDetectionReceiver, filter)
    }
    
    private fun sendEvent(eventName: String, params: WritableMap?) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }
    
    override fun onCatalystInstanceDestroy() {
        super.onCatalystInstanceDestroy()
        fallDetectionReceiver?.let {
            reactApplicationContext.unregisterReceiver(it)
        }
    }
}
